source/agora/test/Base.d

```diff
@@ -99,6 +100,34 @@ shared static this()
     Runtime.extendedModuleUnitTester = &customModuleUnitTester;
 }
 
+void testAssertHandler (string file, ulong line, string msg) nothrow
+{
+    // `std.typecons` test an assert failure, so handle it explicitly
+    static immutable Typecons = "std/typecons.d";
```

Comment:
This is very ugly. What about Windows with backslashes?

Also, you can use `endsWith` from Phobos  in the future.

Reply:
Right. We could just match on `typecons.d`.
By the way I suggested this because otherwise the program would try to print error messages when the unittests for `std.typecons`  were run, resulting in output even when everything was passing.

Reply:
It would be better to explicitly check that it **is** an Agora or SCPD module package, instead of adding special cases for every bad usage of catching AssertError in libraries we cannot easily fix.

Reply:
The problem is that it is triggered from an `agora` package. Originally I went to check that the test runner was only running agora's unittests, and it was! The unittest is apparently called from `agora.common.Serializer`... Which doesn't even import `typecons`...

Reply:
```
[/usr/local/Cellar/ldc/1.23.0/include/dlang/ldc/std/typecons.d:2979] Assertion thrown during test: Called `get' on null Nullable!int.
object.Exception@source/agora/test/Base.d(122): Oops
----------------
??:? nothrow void agora.test.Base.testAssertHandler(immutable(char)[], ulong, immutable(char)[]) [0x107807564]
??:? onAssertErrorMsg [0x10838ea21]
??:? _d_assert_msg [0x10838ee35]
??:? inout pure nothrow ref @property @nogc @safe inout(int) std.typecons.Nullable!(int).Nullable.get() [0x1079146d5]
??:? inout pure nothrow ref @property @nogc @safe inout(int) std.typecons.Nullable!(int).Nullable.get_() [0x107911958]
??:? pure nothrow @nogc @safe int std.typecons.Nullable!(immutable(char)[]).Nullable.__unittest_L2998_C13_1().__dgliteral1() [0x10791193c]
??:? pure void std.exception.assertThrown!(core.exception.AssertError, int).assertThrown(lazy int, immutable(char)[], immutable(char)[], ulong) [0x10791177b]
??:? pure void std.typecons.Nullable!(immutable(char)[]).Nullable.__unittest_L2998_C13_1() [0x1079116b8]
??:? agora.common.Serializer.__unittest [0x107bf97d8]
??:? void agora.test.Base.customModuleUnitTester().runTest(agora.test.Base.customModuleUnitTester().ModTest) [0x107807d46]
??:? core.runtime.UnitTestResult agora.test.Base.customModuleUnitTester() [0x10780712d]
```

This is the stack trace. Although, let me test it on Linux, given OSX's reliable debug infos.

Reply:
I meant something like an inclusion filter like this:

```D
void testAssertHandler (string file, ulong line, string msg) nothrow
{
    if (file.canFind("/agora/") || file.canFind("/scpd/"))
    {
        try
        {
            scope output = stdout.lockingTextWriter();
            output.formattedWrite(
                "================================ ASSERT HANDLER ===============================\n");
            output.formattedWrite!"[%s:%s] Assertion thrown during test: %s\n"
                (file, line, msg);
            CircularAppender!()().print(output);
        }
        catch (Exception exc)
        {
            scope exc_name = typeid(exc).name;
            printf("Could not print thread logs because of %.*s (%.*s:%lu): %.*s\n",
                cast(int) exc_name.length, exc_name.ptr,
                cast(int) file.length, file.ptr, line, cast(int) msg.length, msg.ptr);
        }

        // abort(0);  // do we want this still?
    }

    // TODO: Using `abort` here would give use a core dump,
    // but this gives us a stack trace.
    throw new AssertError(msg, file, line);
}
```

Also plus the `// abort` line I added, maybe we can keep it there if we want?

Reply:
That would not work with assertions triggered in code that we call (e.g. a library / localrest) from our main code.

Reply:
Double checked with Linux, same origin (`Serializer`).

Reply:
Ah so you're saying we still want the circular appender logs in that case? Well if that's so, then we sadly have to keep using this workaround. :'(